walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -35,106 +40,263 @@ public int compare(Event a, Event b) {
+     * ascending order.
+     *
+     * @param minTime minimum length of time for meeting
+     * @param mandatoryAttendees everyone who needs to attend this meeting
+     */
+    private ArrayList<TimeRange> getMandatoryAttendeesMeetingTimes(
+        long minTime, HashSet<String> mandatoryAttendees) {
+      ArrayList<Event> relevantEvents = getRelevantEvents(mandatoryAttendees);
+      ArrayList<TimeRange> possibleMeetingTimes = new ArrayList<TimeRange>();
+
+      // Need to check this so we don't access out of bounds when we add first gap.
+      if (relevantEvents.isEmpty()) {
+        addIfLongEnough(
+            TimeRange.fromStartEnd(0, TimeRange.END_OF_DAY, true), possibleMeetingTimes, minTime);
+        return possibleMeetingTimes;
+      }
+      Collections.sort(relevantEvents, ORDER_BY_START_ASC);
+
+      // Add first gap.
+      addIfLongEnough(
+          TimeRange.fromStartEnd(0, relevantEvents.get(0).getWhen().start(), false),
+          possibleMeetingTimes,
+          minTime);
+      int end = relevantEvents.get(0).getWhen().end();
+      for (Event event : relevantEvents) {
+        // event can be merged with current time range
+        if (event.getWhen().start() <= end) {
+          end = Math.max(end, event.getWhen().end());
+          continue;
         }
+        // Add the time range we were tracking, start a new one from event.
+        addIfLongEnough(
+            TimeRange.fromStartEnd(end, event.getWhen().start(), false),
+            possibleMeetingTimes,
+            minTime);
+        end = event.getWhen().end();
+      }
+
+      // Add the last one we were tracking.
+      addIfLongEnough(
+          TimeRange.fromStartEnd(end, TimeRange.END_OF_DAY, true), possibleMeetingTimes, minTime);
+      return possibleMeetingTimes;
+    }
+
+    /**
+     * Adds range to ranges if it is long enough to fit in a meeting.
+     *
+     * @param range the range being considered
+     * @param ranges the list of ranges >= meetingDuration
+     * @param meetingDuration the duration of meeting to be scheduled
+     */
+    private static void addIfLongEnough(
+        TimeRange range, List<TimeRange> ranges, long meetingDuration) {
+      if (range.duration() >= meetingDuration) {
+        ranges.add(range);
       }
-      if (isRelevant) {
-        relevantEvents.add(event);
+    }
+
+    /**
+     * Returns only those events that are attended by at least one attendee that is attending the
+     * meeting we are trying to schedule. More intuitively, an event is "relevant" if it is attended
+     * by at least one "relevant" attendee.
+     *
+     * @param requestAttendees the set of attendees attending the meeting ("relevant" people)
+     */
+    private ArrayList<Event> getRelevantEvents(HashSet<String> relevantAttendees) {
+      ArrayList<Event> relevantEvents = new ArrayList<Event>();
+      for (Event event : events) {
+        boolean isRelevant = false;
+        for (String person : event.getAttendees()) {
```

Comment:
This can be expressed more tersely using Java8's functional programming paradigm support.

```java
  return events.stream()
     .filter(e -> e.getAttendees().stream().filter(relevantAttendees::contains).findAny())
     .collect(Collectors.toList());

  // N.B. if you _really_ want to return an array list, you need to new ArrayList<T>() the result of that
```

What this does is convert events to a stream (a type of "lazy iterator"), then filters the events where an attendee is in relevantAttendees)